drivers/ethernet/phy/phy_tja1103.c

```diff
@@ -0,0 +1,311 @@
+					 uint16_t reg, uint16_t val)
+{
+	const struct phy_tja1103_config *cfg = dev->config;
+
+	return mdio_write_c45(cfg->mdio, cfg->phy_addr, devad, reg, val);
+}
+
+static inline int phy_tja1103_c45_read(const struct device *dev, uint16_t devad,
+					 uint16_t reg, uint16_t *val)
+{
+	const struct phy_tja1103_config *cfg = dev->config;
+
+	return mdio_read_c45(cfg->mdio, cfg->phy_addr, devad, reg, val);
+}
+
+static int phy_tja1103_reg_read(const struct device *dev, uint16_t reg_addr,
+				 uint32_t *data)
+{
+	const struct phy_tja1103_config *cfg = dev->config;
+	int ret;
+
+	mdio_bus_enable(cfg->mdio);
+
+	ret = phy_tja1103_c22_read(dev, reg_addr, (uint16_t *) data);
+
+	mdio_bus_disable(cfg->mdio);
+
+	return ret;
+}
+
+static int phy_tja1103_reg_write(const struct device *dev, uint16_t reg_addr,
+				  uint32_t data)
+{
+	const struct phy_tja1103_config *cfg = dev->config;
+	int ret;
+
+	mdio_bus_enable(cfg->mdio);
+
+	ret = phy_tja1103_c22_write(dev, reg_addr, (uint16_t) data);
+
+	mdio_bus_disable(cfg->mdio);
+
+	return ret;
+}
+
+static int phy_tja1103_id(const struct device *dev, uint32_t *phy_id)
+{
+	uint16_t val;
+
+	if (phy_tja1103_c22_read(dev, MII_PHYID1R, &val) < 0) {
+		return -EIO;
+	}
+
+	*phy_id = (val & UINT16_MAX) << 16;
+
+	if (phy_tja1103_c22_read(dev, MII_PHYID2R, &val) < 0) {
+		return -EIO;
+	}
+
+	*phy_id |= (val & UINT16_MAX);
+
+	return 0;
+}
+
+static int phy_tja1103_await_phyid(const struct device *dev)
+{
+	int ret;
+	uint32_t count;
+	uint32_t phy_id;
+
+	k_sleep(K_USEC(100000));
+
+	/**
+	 * wait until out of reset.
+	 */
+	for (count = 0U; count < TJA1103_PHY_AWAIT_RETRY_COUNT; ++count) {
+		ret = phy_tja1103_id(dev, &phy_id);
+		if (ret >= 0) {
+			if (phy_id == TJA1103_PHY_ID) {
+				break;
+			}
+			ret = -ETIMEDOUT;
+		}
+		k_sleep(K_USEC(TJA1103_PHY_AWAIT_DELAY_POLL_US));
+	}
+
+	return ret;
+}
+
+static int phy_tja1103_get_link_state(const struct device *dev,
+				       struct phy_link_state *state)
+{
+	struct phy_tja1103_data *const data = dev->data;
+	uint16_t val;
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	if (phy_tja1103_c45_read(dev, MDIO_MMD_PMAPMD,
+				    PMA_STATUS1, &val) >= 0) {
+
+		data->state.is_up = (val & PMA_STATUS_LINK_STATUS) != 0;
+
+		memcpy(state, &data->state, sizeof(struct phy_link_state));
+	}
+
+	k_sem_give(&data->sem);
+
+	return 0;
+}
+
+static int phy_tja1103_cfg_link(const struct device *dev,
+				 enum phy_link_speed adv_speeds)
+{
+	ARG_UNUSED(dev);
+
+	if (!!(adv_speeds & LINK_FULL_100BASE_T)) {
+		return 0;
+	}
+
+	return -ENOTSUP;
+}
+
+static int phy_tja1103_init(const struct device *dev)
+{
+	const struct phy_tja1103_config *const cfg = dev->config;
+	struct phy_tja1103_data *const data = dev->data;
+	uint32_t phy_id;
+	uint16_t val;
+	int ret;
+
+	data->state.is_up = false;
+	data->state.speed = LINK_FULL_100BASE_T;
+
+	ret = phy_tja1103_await_phyid(dev);
+	if (ret < 0) {
+		LOG_ERR("PHY %u didn't come out of reset, %d",
```

Comment:
shouldn't trigger a soft reset before this? see https://elixir.bootlin.com/linux/latest/source/drivers/net/phy/nxp-c45-tja11xx.c#L1305